src/sqlfluff/utils/reflow/elements.py

```diff
@@ -0,0 +1,388 @@
+"""Dataclasses for reflow work."""
+
+
+from itertools import chain
+import logging
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Sequence, Set, Tuple, cast
+
+from sqlfluff.core.parser import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments.raw import WhitespaceSegment
+from sqlfluff.core.rules.base import LintFix
+
+from sqlfluff.utils.reflow.config import ReflowConfig
+from sqlfluff.utils.reflow.depthmap import DepthInfo
+
+# We're in the utils module, but users will expect reflow
+# logs to appear in the context of rules. Hence it's a subset
+# of the rules logger.
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class ReflowElement:
+    """Base reflow element class."""
+
+    segments: Sequence[RawSegment]
+
+    @staticmethod
+    def _class_types(segments: Sequence[RawSegment]) -> Set[str]:
+        return set(chain.from_iterable(seg.class_types for seg in segments))
+
+    @property
+    def class_types(self):
+        """The set of contained class types.
+
+        Parallel to BaseSegment.class_types
+        """
+        return self._class_types(self.segments)
+
+
+@dataclass(frozen=True)
+class ReflowBlock(ReflowElement):
+    """Class for keeping track of elements to reflow.
+
+    It holds segments to reflow and also exposes configuration
+    around how they are expected to reflow around others.
+
+    The attributes exposed are designed to be "post configuration"
+    i.e. they should reflect configuration appropriately.
+
+    NOTE: These are the smallest unit of "work" within
+    the reflow methods, and may contain meta segments.
+    """
+
+    # Options for spacing rules are:
+    # - single:         the default (one single space)
+    # - close:          no whitespace
+    # - close_inline:   no whitespace and no newlines
```

Comment:
How would it look if these were two configs, one for spacing, and one for newlines?

Reply:
So I think there _will_ be two configs. This is what I'm currently working with on my branch were I'm tackling newlines (as an example with commas):

```ini
[sqlfluff:layout:type:comma]
spacing_before = close
line_position = trailing
```

Which I think is what you're implying.

At the moment the only segments which use `close_inline` for their _spacing_ is `dot` - which maybe I should be able to deprecate later, but for now it was important to be able to _prevent_ newlines in that context. You're right though that this feels weird. I might review that part and see if I can get it to work without this option for the config.

Leave this one with me to see what other options might work. Worst case I don't think it's _awful_ if there's a little grey between the two config values, as long as their intent is different.

Reply:
Ok - original `close_inline` was bad idea - agreed. I did that before I had the `spacing_within` config working, and that's a much better fit.

I've migrated over to that and so we now have:

```ini
[sqlfluff:layout:type:object_reference]
spacing_within = inline
```

rather than a spacing config tied to the `dot` as before. Does that feel cleaner @NiallRees ?